Kademlia protocol implemented (#353)
planetarium/libplanet: CHANGES.md

Libplanet changelog
===================

Version 0.6.0
-------------

To be released.

### Backward-incompatible interface changes

 -  `Swarm<T>.AddPeersAsync()` method removed. To connect with seed peers, use
    `Swarm<T>.BootstrapAsync()` method instead.  [[#353]]

### Added interfaces

 -  Added `Swarm<T>.BootstrapAsync()` method to connect with seed peers.  [[#353]]

> 

Reviewer: I think we should introduce `Swarm<T>.PrepareAsync()`.

 -  `StoreExtension.LookupStateReference<T>()` method became to return
    `Tuple<HashDigest<SHA256>, long>` which is a nullable tuple of `Block<T>.Hash`
    and `Block<T>.Index`.  [[#350]]

>     

Reviewer: ```suggestion
```

>  - `Swarm<T>.AddPeersAsync()` method became internal method. To connect with

Reviewer: ```suggestion
 -  `Swarm<T>.AddPeersAsync()` method became internal method. To connect with
```

>     seed peers, use `Swarm<T>.BootstrapAsync()` method instead. [[#353]]

Reviewer: ```suggestion
     seed peers, use `Swarm<T>.BootstrapAsync()` method instead. [[#353]]
```

> 

Reviewer: ```suggestion
```

        `IStore.IterateStagedTransactionIds(bool)` method.
     -  `IStore.StageTransactionIds(IDictionary<TxId, bool>)` method became
        replaced by `StageTransactionIds(IImmutableSet<TxId>()`.
 -  Removed `Swarm<T>.AddPeersAsync()` method. To connect with seed peers, use
    `Swarm<T>.BootstrapAsync()` method instead.  [[#353]]

### Added interfaces

 -  Added `Swarm<T>.PrepareAsync()` method. The method should be called before
    calling `Swarm<T>.BootstrapAsync()`, `Swarm<T>.PreloadAsync()` and
    `Swarm<T>.StartAsync()`.  [[#353]]
 -  Added `Swarm<T>.BootstrapAsync()` method to connect with seed peers.  [[#353]]

### Behavioral changes

 -  NetMQ instances are now initialized at `Swarm<T>.StartAsync()` instead of
    `Swarm<T>()`.  [[#353]]
 -  Peers now connected via [Kademlia protocol][Kademlia]. Peers are now selectively
    connected to each peer.  [[#353]]

>  -  TxIds and Blocks are now broadcasted to selected peers from routing table of

Reviewer 2: ```suggestion
 -  `TxId`s and `Block`s are now broadcasted to selected peers from routing table of
```

 -  Added `StoreExtension.ListAllStateReferences(this IStore, string,
    HashDigest<SHA256>?, HashDigest<SHA256>?)` extension method.
    [[#363], [#384], [#385]]

>     

Reviewer: ```suggestion
```

>  - Added `Swarm<T>.BootstrapAsync()` method to connect with seed peers.  [[#353]]

Reviewer: ```suggestion
 -  Added `Swarm<T>.BootstrapAsync()` method to connect with seed peers.  [[#353]]
```

>     

Reviewer: ```suggestion
```

>  -  Added `StoreExtension.ListAllStateReferences(this IStore, string)` extension

Reviewer: It seems duplicated.

    method.  [[#363]]

>     

Reviewer: ```suggestion
```

 -  `BlockChain<T>.PreloadAsync()` method became to omit rendering of
    `IAction`s in the preloaded behind blocks.  [[#272], [#343]]

>     

Reviewer: ```suggestion
```

 -  `Swarm<T>` became to have two more message types: `GetRecentStates` (`0x0b`)
    and `RecentStates` (`0x0c`).  [[#272], [#343]]

>     

Reviewer: ```suggestion
```

>  - Peers now connected via Kademlia protocol. Peers are now selectively connected to

Reviewer: ```suggestion
 -  Peers now connected via Kademlia protocol. Peers are now selectively connected to
```

>  - Peers now connected via Kademlia protocol. Peers are now selectively connected to

Reviewer: Also, it would be better if there is a link about [Kademlia](https://en.wikipedia.org/wiki/Kademlia).

>     each peer. [[#353]]

Reviewer: ```suggestion
     each peer. [[#353]]
```

> 

Reviewer: ```suggestion
```

> - TxIds and Blocks are now broadcasted to selected peers from routing table of the host

Reviewer: ```suggestion
 -  TxIds and Blocks are now broadcasted to selected peers from routing table of the host
```

>   peer. [[#353]]

Reviewer: ```suggestion
    peer. [[#353]]
```

> 

Reviewer: ```suggestion
```

> - `NetMQPoller` is not used anymore. [[#353]]

Reviewer: Is this needed for libplanet user?

 -  `BlockChain<T>.MineBlock()` and `BlockChain<T>.GetNextTxNonce()` methods
    became to ignore transactions that didn't follow `Transaction<T>.Nonce`
    sequentially and treat them as pendings.  [[#365]]

>     

Reviewer: ```suggestion
```

[#468]: https://github.com/planetarium/libplanet/pull/468

[#353]: https://github.com/planetarium/libplanet/pull/353
[Kademlia]: https://en.wikipedia.org/wiki/Kademlia

> 

Reviewer 2: It seems a wrong place.
